Mathlib/Algebra/Module/CharacterModule.lean

```diff
@@ -0,0 +1,189 @@
+/-
+Copyright (c) 2023 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang, Junyan Xu
+-/
+
+import Mathlib.Algebra.Module.LinearMap.Basic
+import Mathlib.Algebra.Category.ModuleCat.Basic
+import Mathlib.Algebra.Category.GroupCat.Injective
+import Mathlib.Topology.Instances.AddCircle
+import Mathlib.Topology.Instances.Rat
+import Mathlib.LinearAlgebra.Isomorphisms
+import Mathlib.RingTheory.Ideal.Quotient
+import Mathlib.Algebra.Category.GroupCat.EpiMono
+
+/-!
+# Character module of a module
+
+For commutative ring `R` and an `R`-module `M` and an injective module `D`, its character module
+`M⋆` is defined to be `R`-linear maps `M ⟶ D`.
+
+`M⋆` also has an `R`-module structure given by `(r • f) m = f (r • m)`.
+
+## Main results
+
+- `CharacterModuleFunctor` : the contravariant functor of `R`-modules where `M ↦ M⋆` and
+an `R`-linear map `l : M ⟶ N` induces an `R`-linear map `l⋆ : f ↦ f ∘ l` where `f : N⋆`.
+- `LinearMap.dual_surjective_of_injective` : If `l` is injective then `l⋆` is surjective,
+  in another word taking character module as a functor sends monos to epis.
+- `CharacterModule.homEquiv` : there is a bijection between linear map `Hom(N, M⋆)` and
+  `(N ⊗ M)⋆` given by `curry` and `uncurry`.
+
+-/
+
+open CategoryTheory
+
+universe uR uA uB
+
+variable (R : Type uR) [CommRing R]
+variable (A : Type uA) [AddCommGroup A]
+variable (B : Type uB) [AddCommGroup B]
+
+/--
+the character module of abelian group `A` in unit rational circle is `A⋆ := Hom_ℤ(A, ℚ ⧸ ℤ)`
+-/
+def CharacterModule : Type uA := A →+ (AddCircle (1 : ℚ))
+
+namespace CharacterModule
+
+instance : FunLike (CharacterModule A) A (AddCircle (1 : ℚ)) where
+  coe c := c.toFun
+  coe_injective' _ _ _ := by aesop
+
+instance : LinearMapClass (CharacterModule A) ℤ A (AddCircle (1 : ℚ)) where
+  map_add _ _ _ := by rw [AddMonoidHom.map_add]
+  map_smulₛₗ _ _ _ := by rw [AddMonoidHom.map_zsmul, RingHom.id_apply]
+
+instance : AddCommGroup (CharacterModule A) :=
+  inferInstanceAs (AddCommGroup (A →+ _))
+
+section module
+
+variable [Module R A]  [Module R B]
+
+instance : Module R (CharacterModule A) :=
+  Module.compHom (A →+ _) (RingEquiv.toOpposite _ |>.toRingHom : R →+* Rᵈᵐᵃ)
+
+variable {R A B}
+
+@[simp] lemma smul_apply (c : CharacterModule A) (r : R) (a : A) : (r • c) a = c (r • a) := rfl
+
+/--
+Given an abelian group homomorphism `f : A → B`, then `f⋆(L) := L ∘ f` defines a linear map
+between `B⋆` and `A⋆`
+-/
+@[simps] def dual (f : A →ₗ[R] B) : CharacterModule B →ₗ[R] CharacterModule A where
+  toFun L := L.comp f.toAddMonoidHom
+  map_add' := by aesop
+  map_smul' r c := DFunLike.ext _ _ fun x ↦ by
+    simp only [RingHom.id_apply, smul_apply]
+    rw [AddMonoidHom.comp_apply, AddMonoidHom.comp_apply, smul_apply]
+    erw [smul_apply, f.map_smul]
+    rfl
+
+lemma dual_surjective_of_injective (f : A →ₗ[R] B) (hf : Function.Injective f) :
+    Function.Surjective <| dual f := by
```

Comment:
This is defeq to `Module.Baer.extension_property_addMonoidHom` in my modification to #8905. For this reason I'm marking this PR dependent on #8905.

Reply:
thanks

Reply:
Will you incorporate [these golfs](https://github.com/leanprover-community/mathlib4/compare/zjj/CharacterModuleDefsAndBasicAPI...jyxu/CharacterModule) too?

Reply:
Thanks for the golfing. The `$()` trick is amazing, what do they do please.

Reply:
The congr elaborator was added in [#2544](https://github.com/leanprover-community/mathlib4/commit/9a9e31c76cebda72027752554b7da6c991fcc231) and allows you to combine multiple congr_arg and (DFunLike.)congr_fun into a single `congr()` expression; the equality (or equalities) involved need to be labeled with `$`.